File: dotnet/DocumentIntelligence.Api/LlmProviders/OpenAiProvider.cs
using System.Text.Json;
using OpenAI;
using OpenAI.Chat;

namespace DocumentIntelligence.Api.LlmProviders;

public class OpenAiProvider : ILlmProvider
{
    private readonly IConfiguration _config;
    private readonly string _model;

    public string Name => "openai";

    public OpenAiProvider(IConfiguration config, string? model = null)
    {
        _config = config;
        _model = model ?? "gpt-4o";
    }

    public async Task<Dictionary<string, object>> AnalyzeDocumentAsync(
        string? text = null,
        List<byte[]>? images = null,
        byte[]? rawFileBytes = null,
        string? mimeType = null,
        string? extractionHint = null,
        CancellationToken cancellationToken = default)
    {
        var apiKey = _config["LlmSettings:OpenAiApiKey"]
            ?? throw new InvalidOperationException("OpenAI API key not configured");

        var client = new OpenAIClient(apiKey);
        var chatClient = client.GetChatClient(_model);

        var contentParts = new List<ChatMessageContentPart>();

        if (images is { Count: > 0 })
        {
            foreach (var img in images.Take(20))
            {
                var b64 = Convert.ToBase64String(img);
                contentParts.Add(ChatMessageContentPart.CreateImagePart(
                    new BinaryData(img), "image/png", ChatImageDetailLevel.High));
            }
        }

        if (!string.IsNullOrWhiteSpace(text))
        {
            contentParts.Add(ChatMessageContentPart.CreateTextPart($"Document text content:\n\n{text}"));
        }

        if (contentParts.Count == 0)
        {
            return new Dictionary<string, object>
            {
                ["document_type"] = "unknown",
                ["title"] = "Empty Document",
                ["confidence"] = 0.0,
                ["content"] = new Dictionary<string, object>()
            };
        }

        contentParts.Add(ChatMessageContentPart.CreateTextPart("Analyze this document and return structured JSON."));

        var messages = new List<ChatMessage>
        {
            new SystemChatMessage(LlmSystemPrompt.Build(extractionHint)),
            new UserChatMessage(contentParts)
        };

        var options = new ChatCompletionOptions { MaxOutputTokenCount = 8192 };
        var response = await chatClient.CompleteChatAsync(messages, options, cancellationToken);
        var raw = response.Value.Content[0].Text.Trim();

        if (raw.StartsWith("```"))
        {
            raw = raw.Split('\n', 2)[1];
            var lastFence = raw.LastIndexOf("```", StringComparison.Ordinal);
            if (lastFence >= 0)
                raw = raw[..lastFence].Trim();
        }

        return JsonSerializer.Deserialize<Dictionary<string, object>>(raw)
            ?? throw new InvalidOperationException("LLM returned invalid JSON");
    }
}

File: app/llm/openai_provider.py
import base64
import json
from typing import Any

import openai

from app.config import settings
from app.llm.base import LLMProvider


class OpenAIProvider(LLMProvider):
    def __init__(self, api_key: str | None = None, model: str = "gpt-4o"):
        self.client = openai.AsyncOpenAI(api_key=api_key or settings.openai_api_key)
        self.model = model

    async def analyze_document(
        self,
        text: str | None = None,
        images: list[bytes] | None = None,
        extraction_hint: str | None = None,
    ) -> dict[str, Any]:
        content = []

        if images:
            for img in images[:20]:
                b64 = base64.standard_b64encode(img).decode("utf-8")
                content.append(
                    {
                        "type": "image_url",
                        "image_url": {"url": f"data:image/png;base64,{b64}", "detail": "high"},
                    }
                )

        if text:
            content.append({"type": "text", "text": f"Document text content:\n\n{text}"})

        if not content:
            return {
                "document_type": "unknown",
                "title": "Empty Document",
                "confidence": 0.0,
                "content": {},
            }

        content.append(
            {"type": "text", "text": "Analyze this document and return structured JSON."}
        )

        response = await self.client.chat.completions.create(
            model=self.model,
            max_tokens=8192,
            messages=[
                {"role": "system", "content": self._build_system_prompt(extraction_hint)},
                {"role": "user", "content": content},
            ],
        )

        raw = response.choices[0].message.content.strip()
        if raw.startswith("```"):
            raw = raw.split("\n", 1)[1].rsplit("```", 1)[0].strip()

        return json.loads(raw)
